src/spi2.rs

```diff
@@ -0,0 +1,681 @@
+//! SPI peripheral control
+//!
+//! Currently only implements full duplex controller mode support.
+//!
+//! SPI0 is reserved for accessing flash and sram and therefore not usable for other purposes.
+//! SPI1 shares its external pins with SPI0 and therefore has severe restrictions in use.
+//!
+//! SPI2 & 3 can be used freely.
+//!
+//! The CS pin is controlled by hardware on esp32 (contrary to the description of embedded_hal).
+//!
+//! The [Transfer::transfer], [Write::write] and [WriteIter::write_iter] functions lock the
+//! APB frequency and therefore the requests are always run at the requested baudrate.
+//! The primitive [FullDuplex::read] and [FullDuplex::send] do not lock the APB frequency and
+//! therefore may run at a different frequency.
+//!
+//! # TODO
+//! - Quad SPI
+//! - DMA
+//! - Multiple CS pins
+//! - Slave
+
+// TODO -> transfer size & DMA
+// update usage example
+// add update device config Fn
+// &mut self in transaction only needed for sw_cs
+
+use core::cmp::{max, min, Ordering};
+use core::marker::PhantomData;
+use core::ptr;
+
+use std::borrow::Borrow;
```

Comment:
Not OK. Use these from the `alloc` crate. For mutex - use the mutex supplied by `esp-idf-hal` itself. For now, the policy is to avoid `STD` types where possible. it is another question whether this will pay off in any way, but for now we are doing it.

Also - and as a matter of fact, - we also try to avoid the `alloc` module in `esp-idf-hal`, whenever possible.